### PR TITLE
CB-13608 In case of a CM service is in bad health, we should provide …

### PR DESCRIPTION
**Pull request by bbihari**
…cloud storage link for logs

### CODE REVIEW

**Review by lnardai — APPROVED** [review]
LGTM, but one or two line of  logs would be nice in the ServiceStatusCheckerLogLocationDecorator.

**Comment by doktoric** [comment]
lets wait for the logs what @lnardai requested

**Comment by bbihari** [comment]
@doktoric done

File: core/src/test/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecoratorTest.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    @Test
+    void testDecorateWithNoServiceFailure() {
+        when(extendedHostStatuses.hasHostUnhealthyServices(any())).thenReturn(false);
+        Map<InstanceMetadataView, Optional<String>> result = underTest.decorate(getInstanceMetadataWithReason(), extendedHostStatuses, stackDto);
+        assertEquals("DefaultMessage.",
+                result.entrySet().stream().findFirst().flatMap(Map.Entry::getValue).orElse(""));
+        verify(componentConfigProviderService, times(0)).getComponent(anyLong(), any(), any());
+    }
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecoratorTest.java:87]: There seems to be no test case for the `InstanceStatus.SERVICES_UNHEALTHY == instanceMetadataView.getInstanceStatus()` scenario.

**Review by lajosrodek — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecorator.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            String logLocation = getLogLocationFromComponent(stack.getId());
+            String locationMessage = String.format("Please check the logs at the following location: %s", logLocation);
+            for (Map.Entry<InstanceMetadataView, Optional<String>> entry : instancesWithReason.entrySet()) {
+                if (hostStatuses.hasHostUnhealthyServices(hostName(entry.getKey().getDiscoveryFQDN())) && StringUtils.isNotBlank(logLocation)) {
+                    LOGGER.debug(String.format("Extend status reason with cloud storage link for instance: %s.", entry.getKey().getDiscoveryFQDN()));
+                    result.put(entry.getKey(), entry.getValue().map(reason -> String.format("%s %s", reason, locationMessage)));
+                } else {
+                    result.put(entry.getKey(), entry.getValue());
+                }
+            }
```
[the remark this answers is not in the document]
Comment (bbihari) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecorator.java:53]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecorator.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        }
+        return result;
+    }
+
+    private String getLogLocationFromComponent(Long stackId) {
+        String result = "";
+        Component component = componentConfigProviderService
+                .getComponent(stackId, ComponentType.TELEMETRY, ComponentType.TELEMETRY.name());
+        if (component == null) {
+            LOGGER.debug(String.format("Telemetry component is null for stack: %s", stackId));
```
[the remark this answers is not in the document]
Comment (bbihari) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecorator.java:63]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecorator.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            return result;
+        }
+        try {
+            result = component.getAttributes().get(Telemetry.class).getLogging().getStorageLocation();
+        } catch (IOException e) {
+            LOGGER.debug(String.format("Cannot get Telemetry from Components for stack: %s", stackId), e);
```
[the remark this answers is not in the document]
Comment (bbihari) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecorator.java:69]: ok

File: core/src/test/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecoratorTest.java (hunk 1 above)
Comment (bbihari) [reply core/src/test/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecoratorTest.java:87]: good catch!

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecorator.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Inject
+    private ComponentConfigProviderService componentConfigProviderService;
+
+    public Map<InstanceMetadataView, Optional<String>> decorate(Map<InstanceMetadataView, Optional<String>> instancesWithReason,
+            ExtendedHostStatuses hostStatuses, StackDto stack) {
+        Map<InstanceMetadataView, Optional<String>> result = instancesWithReason;
+        boolean extendReason = instancesWithReason.entrySet().stream()
+                .anyMatch(e -> hostStatuses.hasHostUnhealthyServices(hostName(e.getKey().getDiscoveryFQDN()))
+                        && InstanceStatus.SERVICES_UNHEALTHY != e.getKey().getInstanceStatus());
```
[the remark this answers is not in the document]
Comment (bbihari) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecorator.java:40]: In this case we are only interested in new unhealthy nodes. The nodes with SERVICES_UNHEALTHY status will be filtered out later in the status checker job. The main concern here was to limit the number of database queries to the minimal so if not strictly necessary to expand the status message then don't do the telemetry component query.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecorator.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        if (extendReason) {
+            String logLocation = getLogLocationFromComponent(stack.getId());
+            if (StringUtils.isNotBlank(logLocation)) {
+                LOGGER.debug("Extend status reason with cloud storage link for instances with service health issues.");
+                String locationMessage = String.format("Please check the logs at the following location: %s", logLocation);
+                result = new HashMap<>();
+                for (Map.Entry<InstanceMetadataView, Optional<String>> entry : instancesWithReason.entrySet()) {
+                    if (hostStatuses.hasHostUnhealthyServices(hostName(entry.getKey().getDiscoveryFQDN())) && StringUtils.isNotBlank(logLocation)) {
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/ServiceStatusCheckerLogLocationDecorator.java:48]: `StringUtils.isNotBlank(logLocation)` is not needed, it is known to be always `true` inside the outer `if`.

**Review by lajosrodek — CHANGES_REQUESTED** [review]
Only one ask to remove the unnecessary term from the `if` condition.

**Review by lajosrodek — APPROVED** [review]
Thanks, LGTM

**Pull request merged**